common/src/main/java/com/microsoft/identity/common/internal/controllers/NativeAuthMsalController.kt

```diff
@@ -39,17 +86,735 @@ class NativeAuthMsalController : BaseNativeAuthController() {
+        ) {
+            if (!fullCacheRecord.accessToken.isExpired) {
+                setAcquireTokenResult(acquireTokenSilentResult, silentTokenCommandParameters, cacheRecords)
+                val refreshOnCommand =
+                    RefreshOnCommand(parameters, this, PublicApiId.MSAL_REFRESH_ON)
+                CommandDispatcher.submitAndForget(refreshOnCommand)
+            } else {
+                Logger.warn(
+                    TAG,
+                    "Access token is expired. Removing from cache..."
+
+                )
+
+                renewAT(
+                    silentTokenCommandParameters,
+                    acquireTokenSilentResult,
+                    tokenCache,
+                    strategy,
+                    fullCacheRecord
+                )
+            }
+        } else if (accessTokenIsNull(fullCacheRecord) ||
+            refreshTokenIsNull(fullCacheRecord) ||
+            silentTokenCommandParameters.isForceRefresh ||
+            !isRequestAuthorityRealmSameAsATRealm(
+                silentTokenCommandParameters.authority,
+                fullCacheRecord.accessToken
+            ) ||
+            !strategy.validateCachedResult(authScheme, fullCacheRecord)
+        ) {
+            if (!refreshTokenIsNull(fullCacheRecord)) {
+                // No AT found, but the RT checks out, so we'll use it
+                renewAT(
+                    silentTokenCommandParameters,
+                    acquireTokenSilentResult,
+                    tokenCache,
+                    strategy,
+                    fullCacheRecord
+                )
+            } else {
+                val exception = RefreshTokenNotFoundException(
+                    ErrorStrings.NO_TOKENS_FOUND,
+                    "No refresh token was found."
+                )
+                Telemetry.emit(
+                    ApiEndEvent()
+                        .putException(exception)
+                        .putApiId(TelemetryEventStrings.Api.LOCAL_ACQUIRE_TOKEN_SILENT)
+                )
+                throw exception
+            }
+        } else if (fullCacheRecord.accessToken.isExpired) {
+            Logger.warn(
+                TAG,
+                "Access token is expired. Removing from cache..."
+            )
+
+            renewAT(
+                silentTokenCommandParameters,
+                acquireTokenSilentResult,
+                tokenCache,
+                strategy,
+                fullCacheRecord
+            )
+        } else {
+            Logger.verbose(
+                TAG,
+                "Returning silent result"
+            )
+            setAcquireTokenResult(acquireTokenSilentResult, silentTokenCommandParameters, cacheRecords)
+        }
+        Telemetry.emit(
+            ApiEndEvent()
+                .putResult(acquireTokenSilentResult)
+                .putApiId(TelemetryEventStrings.Api.LOCAL_ACQUIRE_TOKEN_SILENT)
+        )
```

Comment:
I think all (most) of this is existing code duplicated here? Instead of doing that could you just invoke the logic on BaseController and re-use that?

Reply:
It's similar, but not the same (e.g. introduction of `RefreshTokenNotFoundException`).